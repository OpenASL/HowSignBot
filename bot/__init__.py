__version__ = "21.12.1"
