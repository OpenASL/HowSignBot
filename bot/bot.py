import asyncio
import itertools
import logging

import disnake
from disnake import ApplicationCommandInteraction
from disnake.ext import commands

from . import settings

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX


intents = disnake.Intents.default()
intents.typing = False
intents.presences = False
intents.bans = False
intents.integrations = False
intents.webhooks = False
intents.invites = False
intents.members = True
intents.messages = True
bot = commands.Bot(
    command_prefix=COMMAND_PREFIX,
    case_insensitive=True,
    owner_id=settings.OWNER_ID,
    intents=intents,
    sync_commands_debug=settings.DEBUG,
    reload=settings.DEBUG,
    test_guilds=settings.TEST_GUILDS or None,
)

PRESENCE_COMMANDS = (
    "/sign",
    "/handshape",
    "/catchphrase | /codenames",
    "/schedule new",
    "/watch2gether",
    "/sentence | /idiom",
    "/feedback",
)
CHANGE_PRESENCE_EVERY = 120  # seconds


async def cycle_presence():
    for command in itertools.cycle(PRESENCE_COMMANDS):
        activity = disnake.Activity(
            name=command,
            type=disnake.ActivityType.watching,
        )
        logger.debug(f"changing presence to show command: {command}")
        await bot.change_presence(activity=activity)
        await asyncio.sleep(CHANGE_PRESENCE_EVERY)


@bot.event
async def on_ready():
    bot.loop.create_task(cycle_presence())


@bot.event
async def on_command_error(ctx, error):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await ctx.send(error)
    else:
        logger.error(f"unhandled exception from command: {ctx.command}", exc_info=error)


@bot.event
async def on_slash_command_error(inter: ApplicationCommandInteraction, error: Exception):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await inter.send(error)
    else:
        logger.error(
            f"unhandled exception from command: {inter.application_command.name!r}",
            exc_info=error,
        )


async def set_default_presence():
    activity = disnake.Activity(
        name="/sign",
        type=disnake.ActivityType.watching,
    )
    await bot.change_presence(activity=activity)
