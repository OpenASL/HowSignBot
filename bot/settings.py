import logging

from databases import DatabaseURL
from environs import Env

env = Env(eager=False)
env.read_env()

DEBUG = env.bool("DEBUG", False)
DATABASE_URL = DatabaseURL(env.str("DATABASE_URL", required=True))
TEST_DATABASE_URL = DATABASE_URL.replace(database="test_" + DATABASE_URL.database)
TESTING = env.bool("TESTING", cast=bool, default=False)
LOG_LEVEL = env.log_level("LOG_LEVEL", logging.INFO)
DISCORD_TOKEN = env.str("DISCORD_TOKEN", required=True)
OWNER_ID = env.int("OWNER_ID", required=True)
SECRET_KEY = env.str("SECRET_KEY", required=True)
COMMAND_PREFIX = env.str("COMMAND_PREFIX", "?")
PARTICIPANT_EMOJI = env.str("PARTICIPANT_EMOJI", default=None)
DAILY_MESSAGE_RANDOM_SEED = env.str("DAILY_MESSAGE_RANDOM_SEED", default=None)
PORT = env.int("PORT", 5000)
TEST_GUILDS = env.list("TEST_GUILDS", default=[], subcast=int)

GOOGLE_PROJECT_ID = env.str("GOOGLE_PROJECT_ID", required=True)
GOOGLE_PRIVATE_KEY = env.str("GOOGLE_PRIVATE_KEY", required=True)
GOOGLE_PRIVATE_KEY_ID = env.str("GOOGLE_PRIVATE_KEY_ID", required=True)
GOOGLE_CLIENT_EMAIL = env.str("GOOGLE_CLIENT_EMAIL", required=True)
GOOGLE_TOKEN_URI = env.str("GOOGLE_TOKEN_URI", "https://oauth2.googleapis.com/token")
TOPICS_SHEET_KEY = env.str("TOPICS_SHEET_KEY", required=True)
FEEDBACK_SHEET_KEY = env.str("FEEDBACK_SHEET_KEY", required=True)

ASLPP_SHEET_KEY = env.str("ASLPP_SHEET_KEY", required=True)
ASLPP_SYNC_TAGS = env.bool("ASLPP_SYNC_TAGS", default=True)
ASLPP_GUILD_ID = env.int("ASLPP_GUILD_ID", default=729838318963130449)
ASLPP_MOD_ROLE_ID = env.int("ASLPP_MOD_ROLE_ID", default=729845246384668723)
ASLPP_INTRODUCTIONS_CHANNEL_ID = env.int("ASLPP_INTRODUCTIONS_CHANNEL_ID", required=False)
ASLPP_ACKNOWLEDGED_RULES_ROLE_ID = env.int(
    "ASLPP_ACKNOWLEDGED_RULES_ROLE_ID", required=False
)
ASLPP_BOT_CHANNEL_ID = env.int("ASLPP_BOT_CHANNEL_ID", required=False)
ASLPP_SKILL_ROLE_IDS = env.list("ASLPP_SKILL_ROLE_IDS", subcast=int)
ASLPP_HEARING_SPECTRUM_ROLE_IDS = env.list("ASLPP_HEARING_SPECTRUM_ROLE_IDS", subcast=int)
ASLPP_AGE_ROLE_IDS = env.list("ASLPP_AGE_ROLE_IDS", subcast=int)
ASLPP_ENABLE_UNMUTE_WARNING = env.bool("ASLPP_ENABLE_UNMUTE_WARNING", True)
ASLPP_INACTIVE_DAYS = env.int("ASLPP_INACTIVE_DAYS", 30)
ASLPP_ZOOM_WATCH_LIST = env.list("ASLPP_ZOOM_WATCH_LIST", default=[], subcast=str)
ASLPP_SURVEY_ID = env.str("ASLPP_SURVEY_ID", default=None)
ASLPP_SURVEY_VANITY_ROLE_ID = env.int(
    "ASLPP_SURVEY_VANITY_ROLE_ID", default=None, required=False
)

# Mapping of Discord user IDs => emails
ZOOM_USERS = env.dict("ZOOM_USERS", subcast_keys=int, required=True)
ZOOM_EMAILS = {email: zoom_id for zoom_id, email in ZOOM_USERS.items()}
ZOOM_JWT = env.str("ZOOM_JWT", required=True)
ZOOM_HOOK_TOKEN = env.str("ZOOM_HOOK_TOKEN", required=True)
ZOOM_REPOST_COOLDOWN = env.int("ZOOM_REPOST_COOLDOWN", 30)
ZZZZOOM_URL = env.str("ZZZZOOM_URL", "https://zzzzoom.us")

WATCH2GETHER_API_KEY = env.str("WATCH2GETHER_API_KEY", required=True)
# When to send practice schedules (in Eastern time)
DAILY_PRACTICE_SEND_TIME = env.time("DAILY_PRACTICE_SEND_TIME", "10:00")

SEND_DEPRECATION_MESSAGES = env.bool("SEND_DEPRECATION_MESSAGES", default=False)

env.seal()
