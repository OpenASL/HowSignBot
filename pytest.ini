[pytest]
filterwarnings =
    ignore:.*@coroutine.*:DeprecationWarning
