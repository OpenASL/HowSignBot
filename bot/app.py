import asyncio
import logging

import aiohttp_cors
from aiohttp import web
from ariadne import graphql
from ariadne.constants import PLAYGROUND_HTML

from . import settings
from .bot import bot
from .database import store
from .graphql.schema import schema
from .utils.extensions import walk_extensions

logger = logging.getLogger(__name__)

# Assign app to bot so that extensions can add routes
bot.app = app = web.Application()  # type: ignore

app.cors = cors = aiohttp_cors.setup(
    app,
    defaults={
        "*": aiohttp_cors.ResourceOptions(
            allow_credentials=False,
            expose_headers="*",
            allow_headers="*",
        )
    },
)


async def ping(_):
    return web.Response(body="", status=200)


def graphql_playground(_):
    return web.Response(text=PLAYGROUND_HTML, status=200, content_type="text/html")


async def graphql_server(request):
    data = await request.json()

    success, result = await graphql(
        schema, data, context_value=request, debug=settings.DEBUG
    )

    status = 200 if success else 400
    return web.json_response(result, status=status)


app.add_routes([web.get("/ping", ping)])
resource = bot.app.router.add_resource("/graphql")  # type: ignore
cors.add(resource.add_route("POST", graphql_server))


async def start_bot():
    try:
        await bot.start(settings.DISCORD_TOKEN)
    finally:
        await bot.close()


async def on_startup(app):
    for ext in walk_extensions():
        bot.load_extension(ext)
    await store.connect()
    app["bot_task"] = asyncio.create_task(start_bot())
    app["bot"] = bot


async def on_shutdown(app):
    app["bot_task"].cancel()
    await app["bot_task"]
    await store.disconnect()


app.on_startup.append(on_startup)
app.on_shutdown.append(on_shutdown)
