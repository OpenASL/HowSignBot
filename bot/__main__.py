import logging

from aiohttp import web

from . import __version__, settings
from .app import app

log_format = "%(asctime)s - %(name)s %(levelname)s: %(message)s"
logging.getLogger("discord").setLevel(logging.WARNING)
logging.basicConfig(
    format=log_format,
    level=settings.LOG_LEVEL,
)
logger = logging.getLogger(__name__)

logger.info(f"starting bot version {__version__}")
web.run_app(app, port=settings.PORT)
