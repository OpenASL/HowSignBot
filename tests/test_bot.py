import datetime as dt
import random
from unittest import mock

import gspread
import pytest
from freezegun import freeze_time
from syrupy.filters import props

import bot

random.seed(1)


@pytest.mark.parametrize(
    "word",
    (
        "tiger",
        "||tiger||",
        "what's up",
        "||what's up||",
        "need, ask",
        "||need, ask||",
    ),
)
def test_sign(snapshot, word):
    result = bot.sign_impl(word)
    assert result == snapshot


@pytest.mark.parametrize("name", ("random", "open8", "open9"))
def test_handshape(snapshot, name):
    result = bot.handshape_impl(name)
    assert result == snapshot(
        exclude=props(
            "fp",
        )
    )


def test_handshapes(snapshot):
    result = bot.handshapes_impl()
    assert result == snapshot


def test_catchphrase(snapshot):
    result = bot.catchphrase_impl()
    assert result == snapshot


@pytest.mark.parametrize("category", ("categories", "food", "Hard"))
def test_catchphrase_categories(snapshot, category):
    result = bot.catchphrase_impl(category)
    assert result == snapshot


@pytest.mark.parametrize("spoiler", (None, "||"))
def test_sentence(snapshot, spoiler):
    result = bot.sentence_impl(spoiler)
    assert result == snapshot


@pytest.mark.parametrize("spoiler", (None, "||"))
def test_idiom(snapshot, spoiler):
    result = bot.idiom_impl(spoiler)
    assert result == snapshot


@pytest.fixture
def mock_worksheet(monkeypatch):
    monkeypatch.setattr(bot, "SCHEDULE_SHEET_KEYS", {1234: "abc"}, raising=True)
    monkeypatch.setattr(bot, "GOOGLE_PRIVATE_KEY", "fake", raising=True)
    with mock.patch("bot.get_practice_worksheet_for_guild") as mock_get_worksheet:
        WorksheetMock = mock.Mock(spec=gspread.Worksheet)
        WorksheetMock.get_all_values.return_value = [
            ["docs", "more docs"],
            ["Start time", "Host (optional)", "Notes (optional)"],
            ["Friday 5pm EDT", "Steve", "recurring"],
            ["Wed 6pm edt", "", "another recurring"],
            ["9/26 2pm PDT 2020", "Steve", "one time"],
            ["Sunday, September 27 02:00 PM EDT 2020", "", "another 1Ô∏è‚É£"],
        ]
        mock_get_worksheet.return_value = WorksheetMock
        yield WorksheetMock


@pytest.mark.parametrize(
    "when",
    (
        None,
        "today",
        "tomorrow",
        "saturday",
        "9/27",
    ),
)
@freeze_time("2020-09-25 14:00:00")
def test_schedule(snapshot, monkeypatch, mock_worksheet, when):
    result = bot.schedule_impl(1234, when)
    assert result == snapshot


@pytest.mark.parametrize(
    "start_time",
    (
        "2pm edt",
        "at 2pm edt",
        '2pm edt "chat and games! üéâ"',
        "9/24 1:45pm edt ‚Äúaround 45 min.-1 hour‚Äù",
        "tomorrow 2pm pdt",
        "sunday 10:30pm pdt",
        "9/27 9am cdt",
        '"games" 10am edt Sunday',
        '"games" at 10am edt on Sunday',
        '"games" at 10am edt on 9/27',
        '10am edt "games" on 9/27',
        '"classifiers" at 6pm pdt',
    ),
)
@freeze_time("2020-09-25 14:00:00")
def test_practice(snapshot, monkeypatch, mock_worksheet, start_time):
    bot.practice_impl(guild_id=1234, host="Steve", start_time=start_time)
    mock_worksheet.append_row.assert_called_once()
    assert mock_worksheet.append_row.call_args[0][0] == snapshot


@pytest.mark.parametrize(
    "start_time",
    (
        "invalid",
        "today",
        "tomorrow",
        "today edt",
        "today pdt",
    ),
)
@freeze_time("2020-09-25 14:00:00")
def test_practice_invalid(snapshot, monkeypatch, mock_worksheet, start_time):
    result = bot.practice_impl(guild_id=1234, host="Steve", start_time=start_time)
    mock_worksheet.append_row.assert_not_called()
    assert result == snapshot


@pytest.mark.parametrize(
    ("value", "expected"),
    (
        ('today 2pm "chat"', ("today 2pm", "chat")),
        ('"chat" today 2pm', ("today 2pm", "chat")),
        ("today 2pm", ("today 2pm", None)),
        ('today 2pm ""', ("today 2pm", "")),
        ('today 2pm "steve\'s practice"', ("today 2pm", "steve's practice")),
        ("today 2pm ‚Äúsmart quotes üòû‚Äù", ("today 2pm", "smart quotes üòû")),
    ),
)
def test_get_and_strip_quoted_text(value, expected):
    assert bot.get_and_strip_quoted_text(value) == expected


@pytest.mark.parametrize(
    "value",
    (
        "today 8:25pm edt",
        "today 2pm edt",
        "today at 2pm edt",
        "tomorrow 2pm pdt",
        "on friday at 2pm cst",
        "9/25 2:30 pm cdt",
    ),
)
def test_parse_human_readable_datetime(value):
    dtime = bot.parse_human_readable_datetime(value)
    assert dtime.tzinfo == dt.timezone.utc
