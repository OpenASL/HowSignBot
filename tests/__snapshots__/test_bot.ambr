# name: test_catchphrase
  <class 'dict'> {
    'content': '
      ||skirt||
      ||green||
      ||bay||
      ||snowman||
      ||Steve Martin||
      ||river rafting||
      ||Spain||
      ||dredge||
      Enter `!cp` or `!cp [category]` for more.
    ',
  }
---
# name: test_catchphrase_categories[Hard]
  <class 'dict'> {
    'content': '
      ||wreath||
      ||through||
      ||highway||
      ||cardboard||
      ||full moon||
      ||mast||
      ||wool||
      ||blimp||
      Enter `!cp` or `!cp [category]` for more.
    ',
  }
---
# name: test_catchphrase_categories[categories]
  <class 'dict'> {
    'content': '
      easy, medium, hard, animals, food, travel, people, household
      Enter `!cp` or `!cp [category]` to generate a list of words/phrases.
    ',
  }
---
# name: test_catchphrase_categories[food]
  <class 'dict'> {
    'content': '
      ||scone||
      ||chicken pot pie||
      ||blanch||
      ||clams||
      ||Kool-aid||
      ||corn on the cob||
      ||pineapple||
      ||yam||
      Enter `!cp` or `!cp [category]` for more.
    ',
  }
---
# name: test_handshape[open8]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description=Embed.Empty,
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(url='attachment://Open8.png'),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title='Open8',
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
    'file': <class 'File'> {
      filename='Open8.png',
    },
  }
---
# name: test_handshape[open9]
  <class 'dict'> {
    'content': '"open9" not found. Did you mean "OpenX"? Enter `!handshapes` to see a list of handshapes.',
  }
---
# name: test_handshape[random]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description=Embed.Empty,
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(url='attachment://8.png'),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title='8',
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
    'file': <class 'File'> {
      filename='8.png',
    },
  }
---
# name: test_handshapes
  <class 'dict'> {
    'content': '
      1, 3, Bent3, 4, 5, Claw5, 6, 7, 8, Open8, 9, Flat9, A, OpenA, B, BentB, FlatB, OpenB, C, FlatC, SmallC, D, E, G, H, I, K, L, M, OpenM, N, OpenN, O, FlatO, SmallO, R, S, T, V, BentV, X, OpenX, Y, ILY, Corna
      
      Enter `!handshape` to display a random handshape or `!handshape [name]` to display a specific handshape.
  
    ',
  }
---
# name: test_howsign[tiger]
  <class 'dict'> {
    'content': '
      TIGER
      _Lifeprint_ : https://www.google.com/search?&q=site%3Alifeprint.com+tiger
      _YouGlish_: https://youglish.com/pronounce/tiger/signlanguage/asl
      _Spread The Sign_: <https://www.spreadthesign.com/en.us/search/?q=tiger>
      _SigningSavvy_: https://www.signingsavvy.com/search/tiger
  
    ',
  }
---
# name: test_howsign[what's up]
  <class 'dict'> {
    'content': '
      WHAT'S UP
      _Lifeprint_ : https://www.google.com/search?&q=site%3Alifeprint.com+what%27s+up
      _YouGlish_: https://youglish.com/pronounce/what%27s+up/signlanguage/asl
      _Spread The Sign_: <https://www.spreadthesign.com/en.us/search/?q=what%27s+up>
      _SigningSavvy_: https://www.signingsavvy.com/search/what%27s+up
  
    ',
  }
---
# name: test_howsign[||tiger||]
  <class 'dict'> {
    'content': '
      ||TIGER||
      _Lifeprint_ : || https://www.google.com/search?&q=site%3Alifeprint.com+tiger ||
      _YouGlish_: || https://youglish.com/pronounce/tiger/signlanguage/asl ||
      _Spread The Sign_: || <https://www.spreadthesign.com/en.us/search/?q=tiger> ||
      _SigningSavvy_: || <https://www.signingsavvy.com/search/tiger> ||
  
    ',
  }
---
